src/braidset.js

```diff
@@ -0,0 +1,116 @@
+/**
+ * This module provides functions for BraidSets, which is a set of Braids
+ * There are generally two Braid objects in a BraidSet, one for deposit
+ * and one for change.
+ *
+ * @module braidset
+ */
+
+import { Struct } from "bufio";
+import assert from "assert";
+import { NETWORKS } from "./networks";
+import {MULTISIG_ADDRESS_TYPES} from './multisig';
+import {validateExtendedPublicKey} from './keys';
+import {Braid, deriveMultisigByIndex, deriveMultisigByPath} from './braid';
+import {hardenedBIP32Index} from './paths';
+
+export class BraidSet extends Struct {
+  constructor(options) {
+    super();
+    if (!options || !Object.keys(options).length) {
+      return this;
+    }
+
+    assert(
+      MULTISIG_ADDRESS_TYPES.includes(options.network),
+      `Expected addressType to be one of:  ${MULTISIG_ADDRESS_TYPES}.`
+    );
+    this.addressType = options.addressType;
+    assert(
+      NETWORKS.includes(options.network),
+      `Expected network to be one of:  ${NETWORKS}.`
+    );
+    this.network = options.network;
+
+    options.extendedPublicKeys.forEach((xpub) => {
+      const xpubValidationError = validateExtendedPublicKey(xpub);
+      assert(!xpubValidationError.length, xpubValidationError);
+    })
+    this.extendedPublicKeys = options.extendedPublicKeys;
+
+    assert(typeof options.requiredSigners === "number");
+    assert(this.requiredSigners <= this.extendedPublicKeys.length,
+      `Can't have more requiredSigners than there are keys.`)
+    this.requiredSigners = options.requiredSigners;
+
+    this.depositBraid = new Braid({
+      network: this.network,
+      addressType: this.addressType,
+      extendedPublicKeys: this.extendedPublicKeys,
+      requiredSigners: this.requiredSigners,
+      bip32Path: options.hardened ? hardenedBIP32Index('0') : '0',
+    });
+    this.changeBraid = new Braid({
+      network: this.network,
+      addressType: this.addressType,
+      extendedPublicKeys: this.extendedPublicKeys,
+      requiredSigners: this.requiredSigners,
+      bip32Path: options.hardened ? hardenedBIP32Index('1') : '1',
+    });
```

Comment:
Another thought that came from latest naming discussion, what if the braidSet didn't care about the names of its various braid components, but rather just tracked them by index. So you could create a set with 3 braids, and the consumer of the braid set could just say "get me an address from braid at index x". The braid set shouldn't have to care what the purpose of the braid is for, only the index (although, "purpose" could be a good label to keep track if we wanted).

Reply:
Yep - your thinking aligns here, but based on the general nature of it - index wouldnt necessarily be unique across your braids. It might be, but it migh tnot be since we keep the xpubs defined on the braid and not the braidset. moving them up to the braidset would add that constraint. In any case we're tabling braidset for a bit but this is good stuff to keep the juices/ideas flowing.